### DISCUSSION TITLE
搭建大型源码阅读环境——使用 OpenGrok — 码志

### DISCUSSION THREAD

**Discussion by mzlogin** [mzlogin/blog-comments#206] [Announcements]
https://mazhuang.org/2016/12/14/rtfsc-with-opengrok/

**Comment by beyond2002** [comment]
很详细的指导！
不过在启动 Tomcat时遇到错误提示“The CATALINA_HOME environment variable is not defined correctly This environment variable is needed to run this program”

**Comment by mzlogin** [comment]
@beyond2002 感谢指出，已经将添加 JRE_HOME 和 CATELINA_HOME 环境变量的说明加到文章里了。

**Comment by chongzhe** [comment]
楼主有没有尝试过这款让你RTFSC得更愉快的产品:

https://insight.io/github.com/daimajia/AndroidSwipeLayout/tree/master/

**Comment by mzlogin** [comment]
@chongzhe 还真没有……回头好好试用一下再来评价~ :)

**Comment by mzlogin** [comment]
@chongzhe 

挺赞的！

有几点问题提一下：

1. 每次点击首页的“EXPLORE IN DEMO”都会弹出 onboarding guide，每次要点击很多个 next

2. 一些最想看到的项目，比如 AOSP 最新源码，看不到

3. 没有找到项目内的文件、符号全局搜索的有效方式，也许是我的使用方式不对，不过也没有很容易地找到使用方式介绍

**Comment by chongzhe** [comment]
1. 这个功能主要是给第一次使用的user的，后面其实就不需要再点了，login之后直接开始搜索代码
2. ASOP代码是有的，这两天有个改动的原因暂时unavailable，明天把链接发给你
3. 文件/符号搜索点击搜索按钮或者cmd+shift+s就可以打开，如下图
![image](https://user-images.githubusercontent.com/2304183/28097792-2d8af7f8-6667-11e7-9aad-9e9b57808fa0.png)

**Comment by mzlogin** [comment]
@chongzhe 
 
赞！文件/符号搜索我是在已有的 <https://insight.io/github.com/lambdalab-mirror/android-4.4.2_r1/tree/master/> 里尝试的，我后来才发现它还是 Analyzing 状态，在其它 Indexed 状态的项目里体验很不错。

另外，代码的字体大小可以修改不？

**Comment by chongzhe** [comment]
是的，之前我们是push了一个镜像因为AOSP代码不在github上，现在google在github上也官方提供了镜像如下地址，不就之后我们会把我们自己的镜像sunset掉：
https://insight.io/github.com/android/platform_frameworks_base/tree/master

字体暂时还不能改，之后如果这方面需求多我们可以加上界面的自定义功能。当然，既然是个web应用，如果实在觉得字体大小不舒服，也可以暂时先用cmd + '+'和cmd + '-'调字体...

**Comment by mzlogin** [comment]
@chongzhe 

👍 👍 👍 

多谢！这个已经能满足我大部分时候要查 AOSP 源码的需求了。

能不能 Index 一个完整的 AOSP？上面的项目基本够用了，但有时候看一些 frameworks 之外的，比如系统 App 的源码，这里头还是没有的。

你们的产品挺好用的，可以挑一些大家最常用的项目，做一些精品，然后让更多的人知道，当然也许你们已经在这么做了。:)

另外，有计划支持注释里的 `@link` 和 `@see` 跳转吗？ 😄

**Comment by chongzhe** [comment]
@mzlogin 谢谢建议！

对目前只有frameworks代码，主要是整个AOSP代码使用repo而不是git管理的，所以实现起来有一些难度，我们会在这两个月尽力收录全部AOSP代码。

我们目前有一些Apache项目集和优质Android项目，后续可以做成专题通过博客的形式让更多人知道，关于运营博客可能还有更多的经验从你这边学习:)

@link和@see的支持正好这个sprint正在做，应该下一版就会有。

另外我往你的gmail邮箱里发了我的个人联系方式，希望能随时保持联系！

**Comment by mzlogin** [comment]
@chongzhe 已经申请加你微信。

**Comment by hoohack** [comment]
将 OpenGrok 的 lib 目录里的 source.war 解压到 D:\Programs\apache-tomcat-8.5.8\webapps\source，配置 WEB-INF\web.xml 文件的 CONFIGURATION 为上一步生成的 data 目录下的 configureation.xml

这一步的configureation.xml文件从哪里来呢？

**Comment by mzlogin** [comment]
@hoohack 这里是指定一个文件路径，实际的 configuration.xml 会在第 8 步，建立索引的时候生成。

**Comment by hoohack** [comment]
好的，我看到了。
现在可以运行，但是当我点击索引出来的链接，想查看详细文件的时候却报错了，报的是多次重定向，您有遇到过吗？详细的截图我发你gmail邮箱了

**Comment by mzlogin** [comment]
@hoohack 我换了几个 OpenGrok 的版本试了一下，在昨天使用的 1.1-rc12 这个版本也遇到了，OpenGrok 已经发布了 1.1-rc13，修复了这个问题。详见 https://github.com/OpenGrok/OpenGrok/releases 的 release notes。你换到最新版本试试。

**Comment by hoohack** [comment]
好的，感谢@mzlogin

**Comment by mzlogin** [comment]
@hoohack 不客气 😄

**Comment by hoohack** [comment]
11,12,13，我试了这三个版本的，都不可以呢，你试过这三个版本会吗？ @mzlogin

**Comment by mzlogin** [comment]
我用 13 没问题。记得换版本的时候也要把 tomcat 下的 webapps/source 换掉。 @hoohack

**Comment by hoohack** [comment]
对，把 tomcat 下的 webapps/source 换掉，可以了，非常感谢 @mzlogin

**Comment by mzlogin** [comment]
不客气~都是踩坑踩出来的经验 😄 @hoohack

**Comment by kaoer** [comment]
我配置了多项目，dataRoot设置成代码对应的目录，打开检索出来的文件，进入查看源码时，点击函数，会自动跳转到搜索这个函数的界面，但是每次都会报错，自动搜索的是/source/目录，报错CONFIGURATION parameter has not been configured in web.xml! Please configure your webapp.
dataRoot parameter is not set in configuration.xml!

楼主有没有遇到过？怎么解决的呢？

**Comment by mzlogin** [comment]
> @kaoer
> 我配置了多项目，dataRoot设置成代码对应的目录，打开检索出来的文件，进入查看源码时，点击函数，会自动跳转到搜索这个函数的界面，但是每次都会报错，自动搜索的是/source/目录，报错CONFIGURATION parameter has not been configured in web.xml! Please configure your webapp.
> dataRoot parameter is not set in configuration.xml!
> 
> 楼主有没有遇到过？怎么解决的呢？

把 data root 与 source root 区分一下比较好。

另外如果一直报错，将 OpenGrok 相关的东西都删掉，包括 tomcat 下的 source.war 和 source 目录，然后重新配置试试。

**Comment by kaoer** [comment]
> @mzlogin
> > @kaoer
> > 我配置了多项目，dataRoot设置成代码对应的目录，打开检索出来的文件，进入查看源码时，点击函数，会自动跳转到搜索这个函数的界面，但是每次都会报错，自动搜索的是/source/目录，报错CONFIGURATION parameter has not been configured in web.xml! Please configure your webapp.
> > dataRoot parameter is not set in configuration.xml!
> > 
> > 楼主有没有遇到过？怎么解决的呢？
> 
> 把 data root 与 source root 区分一下比较好。
> 
> 另外如果一直报错，将 OpenGrok 相关的东西都删掉，包括 tomcat 下的 source.war 和 source 目录，然后重新配置试试。

data root 与 source root 是分开的，楼主使用过这个功能吗？我是最近才发现有这个功能。这两天一直在解这个问题。像source insight 一样，可以直接跳转。

**Comment by mzlogin** [comment]
> @kaoer
> > ...
> 
> data root 与 source root 是分开的，楼主使用过这个功能吗？我是最近才发现有这个功能。这两天一直在解这个问题。像source insight 一样，可以直接跳转。

当然用过啦~这是代码阅读工具最基本的功能啊。你不妨把你的配置贴出来看看吧，

source root，data root，web.xml 里的 CONFIGURATION 等等。

**Comment by kaoer** [comment]
> @mzlogin
> > @kaoer
> > > ...
> > 
> > data root 与 source root 是分开的，楼主使用过这个功能吗？我是最近才发现有这个功能。这两天一直在解这个问题。像source insight 一样，可以直接跳转。
> 
> 当然用过啦~这是代码阅读工具最基本的功能啊。你不妨把你的配置贴出来看看吧，
> 
> source root，data root，web.xml 里的 CONFIGURATION 等等。



  <void property="dataRoot">
   <string>/var/data/opengrok/sdm660/data</string>
  </void>

  <void property="sourceRoot">
   <string>/var/data/sdm660</string>
  </void>

    <param-name>CONFIGURATION</param-name>
    <param-value>/var/data/opengrok/sdm660/etc/configuration.xml</param-value>

**Comment by kaoer** [comment]
原格式贴不了，就这样写吧：

dataRoot     /var/data/opengrok/sdm660/data
sourceRoot       /var/data/sdm660
CONFIGURATION    /var/data/opengrok/sdm660/etc/configuration.xml

**Comment by mzlogin** [comment]
> @kaoer
> 原格式贴不了，就这样写吧：
> 
> dataRoot     /var/data/opengrok/sdm660/data
> sourceRoot       /var/data/sdm660
> CONFIGURATION    /var/data/opengrok/sdm660/etc/configuration.xml
> 
> 
> 

如果你这些路径都是真实存在的且符合你的预期，那单从配置上看不出什么问题……尴尬……

更改 source.war 和 tomcat 下的 source 内容时先停止 tomcat，更新完之后再启动？如果你就是这样的操作的那我暂时也想不到是什么原因导致的了。
